ssl/statem/statem_srvr.c

```diff
@@ -1566,6 +1566,68 @@ static int tls_early_post_process_client_hello(SSL *s, int *pal)
 
     s->hit = 0;
 
+    if (!ssl_cache_cipherlist(s, &clienthello->ciphersuites,
+                              clienthello->isv2, &al) ||
+        !bytes_to_cipher_list(s, &clienthello->ciphersuites, &ciphers, &scsvs,
+                             clienthello->isv2, &al)) {
+        goto err;
+    }
+
+    s->s3->send_connection_binding = 0;
+    /* Check what signalling cipher-suite values were received. */
+    if (scsvs != NULL) {
+        for(i = 0; i < sk_SSL_CIPHER_num(scsvs); i++) {
+            c = sk_SSL_CIPHER_value(scsvs, i);
+            if (SSL_CIPHER_get_id(c) == SSL3_CK_SCSV) {
+                if (s->renegotiate) {
+                    /* SCSV is fatal if renegotiating */
+                    SSLerr(SSL_F_TLS_EARLY_POST_PROCESS_CLIENT_HELLO,
+                           SSL_R_SCSV_RECEIVED_WHEN_RENEGOTIATING);
+                    al = SSL_AD_HANDSHAKE_FAILURE;
+                    goto err;
+                }
+                s->s3->send_connection_binding = 1;
+            } else if (SSL_CIPHER_get_id(c) == SSL3_CK_FALLBACK_SCSV &&
+                       !ssl_check_version_downgrade(s)) {
+                /*
+                 * This SCSV indicates that the client previously tried
+                 * a higher version.  We should fail if the current version
+                 * is an unexpected downgrade, as that indicates that the first
+                 * connection may have been tampered with in order to trigger
+                 * an insecure downgrade.
+                 */
+                SSLerr(SSL_F_TLS_EARLY_POST_PROCESS_CLIENT_HELLO,
+                       SSL_R_INAPPROPRIATE_FALLBACK);
+                al = SSL_AD_INAPPROPRIATE_FALLBACK;
+                goto err;
+            }
+        }
+    }
+
+    /* For TLSv1.3 we must select the ciphersuite *before* session resumption */
+    if (SSL_IS_TLS13(s)) {
```

Comment:
We did this in BoringSSL. It's a little trickier than merely removing this check.

The events that end up happening in the old order are:

1. Decide on resumption.
2. Parse extensions (call SNI callback).
3. Call `cert_cb`.
4. Select cipher suite.

The goal is to order (4) before (1), but merely moving (4) doesn't quite work. (4) needs to happen after (2) and (3) because, until the certificate is fixed, you cannot select a TLS 1.2 cipher. You don't know whether to filter out the RSA or ECDSA cipher suite. One of the goals in redoing TLS 1.3 cipher negotiation was precisely to avoid this interaction. The consumer may also have different cipher configurations per SNI or whatever, which also imposes this constraint. Whether you wish to support the latter is still open since TLS 1.3 didn't exist in 1.1.0.

In BoringSSL, we reordered to (2), (3), (4), (1) with this patch series:
* https://boringssl.googlesource.com/boringssl.git/+/4eb95ccfd64d46fad23f36238690594d02518467 (ignore the ALPN note in the commit message; I'd forgotten about [this change](https://boringssl.googlesource.com/boringssl/+/9ef31f01af6c59efe65fe0c4ab3be7316046ccca))
* https://boringssl.googlesource.com/boringssl.git/+/34202b93b6195e9c55402a6bc2653956d0cd26d8
* https://boringssl.googlesource.com/boringssl.git/+/f01f42a2ce161657c6413ef71d7c04262c7abd56

But it had some user-visible effects (I documented the ones I could think of in the commit messages) and we had to patch one consumer. They previously used `SSL_get_session` + `session->tlsext_hostname` which we switched to `SSL_get_servername`. But the end result is much simpler.

Hope this is helpful!